File: src-tauri/src/commands/messages.rs
//! 消息与生成控制命令。
//!
//! 该文件承接消息查询、版本管理与 AI 生成相关的所有 Tauri 命令。
//! 由于生成过程需要通过 Tauri Channel 持续回推事件，这里把 `Channel<GenerationEvent>`
//! 直接作为命令参数暴露给前端。

use tauri::{State, ipc::Channel};

use crate::{
    error::AppError,
    models::{
        DeleteVersionResult, GenerationEvent, MessageNode, RerollInput, RerollResult,
        SendMessageInput, SendMessageResponse, VersionContent,
    },
    services::message_service,
    state::AppState,
};

/// 查询会话下的消息列表。
pub async fn list_messages_impl(
    state: &AppState,
    id: String,
    before_order_key: Option<String>,
    limit: Option<i64>,
) -> Result<Vec<MessageNode>, AppError> {
    message_service::list_messages(state, &id, before_order_key, limit).await
}

/// Tauri 命令：查询会话下的消息列表。
#[tauri::command]
pub async fn list_messages(
    state: State<'_, AppState>,
    id: String,
    before_order_key: Option<String>,
    limit: Option<i64>,
) -> Result<Vec<MessageNode>, AppError> {
    list_messages_impl(state.inner(), id, before_order_key, limit).await
}

/// 按需加载某个版本的完整内容。
pub async fn get_version_content_impl(
    state: &AppState,
    version_id: String,
) -> Result<VersionContent, AppError> {
    message_service::get_version_content(state, &version_id).await
}

/// Tauri 命令：按需加载某个版本的完整内容。
#[tauri::command]
pub async fn get_version_content(
    state: State<'_, AppState>,
    version_id: String,
) -> Result<VersionContent, AppError> {
    get_version_content_impl(state.inner(), version_id).await
}

/// 切换某个楼层的 active version。
pub async fn set_active_version_impl(
    state: &AppState,
    id: String,
    node_id: String,
    version_id: String,
) -> Result<(), AppError> {
    message_service::set_active_version(state, &id, &node_id, &version_id).await
}

/// Tauri 命令：切换某个楼层的 active version。
#[tauri::command]
pub async fn set_active_version(
    state: State<'_, AppState>,
    id: String,
    node_id: String,
    version_id: String,
) -> Result<(), AppError> {
    set_active_version_impl(state.inner(), id, node_id, version_id).await
}

/// 删除指定版本。
pub async fn delete_version_impl(
    state: &AppState,
    id: String,
    node_id: String,
    version_id: String,
) -> Result<DeleteVersionResult, AppError> {
    message_service::delete_version(state, &id, &node_id, &version_id).await
}

/// Tauri 命令：删除指定版本。
#[tauri::command]
pub async fn delete_version(
    state: State<'_, AppState>,
    id: String,
    node_id: String,
    version_id: String,
) -> Result<DeleteVersionResult, AppError> {
    delete_version_impl(state.inner(), id, node_id, version_id).await
}

/// 发送消息并触发后台生成。
pub async fn send_message_impl(
    state: &AppState,
    id: String,
    input: SendMessageInput,
    event_channel: Option<Channel<GenerationEvent>>,
) -> Result<SendMessageResponse, AppError> {
    message_service::send_message(state, &id, input, event_channel).await
}

/// Tauri 命令：发送消息并触发后台生成。
#[tauri::command]
pub async fn send_message(
    state: State<'_, AppState>,
    id: String,
    input: SendMessageInput,
    event_channel: Channel<GenerationEvent>,
) -> Result<SendMessageResponse, AppError> {
    send_message_impl(state.inner(), id, input, Some(event_channel)).await
}

/// 对指定楼层执行 reroll。
pub async fn reroll_impl(
    state: &AppState,
    id: String,
    node_id: String,
    input: Option<RerollInput>,
    event_channel: Option<Channel<GenerationEvent>>,
) -> Result<RerollResult, AppError> {
    message_service::reroll(
        state,
        &id,
        &node_id,
        input.unwrap_or_default(),
        event_channel,
    )
    .await
}

/// Tauri 命令：对指定楼层执行 reroll。
#[tauri::command]
pub async fn reroll(
    state: State<'_, AppState>,
    id: String,
    node_id: String,
    input: Option<RerollInput>,
    event_channel: Channel<GenerationEvent>,
) -> Result<RerollResult, AppError> {
    reroll_impl(state.inner(), id, node_id, input, Some(event_channel)).await
}

/// 取消某个 generating version。
pub async fn cancel_generation_impl(
    state: &AppState,
    version_id: String,
) -> Result<(), AppError> {
    message_service::cancel_generation(state, &version_id).await
}

/// Tauri 命令：取消某个 generating version。
#[tauri::command]
pub async fn cancel_generation(
    state: State<'_, AppState>,
    version_id: String,
) -> Result<(), AppError> {
    cancel_generation_impl(state.inner(), version_id).await
}

File: src/lib/transport/messages.ts
/**
 * 消息查询、版本控制与生成事件相关的 Tauri transport 封装。
 *
 * 这个入口文件只负责：
 * 1. 发起 Tauri `invoke()` 调用。
 * 2. 复用 codec 将 snake_case 载荷转为前端结构。
 * 3. 统一 re-export 消息相关公共类型，避免组件依赖内部拆分文件。
 */

import { invoke } from "@tauri-apps/api/core";
import { toAppError, toOptionalValue, type AppError } from "./common";
import {
  createGenerationChannel,
  fromRawDeleteVersionResult,
  fromRawMessageNode,
  fromRawRerollResult,
  fromRawSendMessageResponse,
  fromRawVersionContent,
  toRawRerollInput,
  toRawSendMessageInput
} from "./message-codecs";
import type {
  DeleteVersionResult,
  GenerationEvent,
  MessageNode,
  RawDeleteVersionResult,
  RawDryRunResult,
  RawMessageNode,
  RawRerollResult,
  RawStartedResult,
  RawVersionContent
} from "./message-types";
import type {
  RerollInput,
  RerollResult,
  SendMessageInput,
  SendMessageResponse,
  VersionContent
} from "./message-types";

/**
 * 获取会话消息列表。
 */
export async function listMessages(
  id: string,
  beforeOrderKey?: string | null,
  limit?: number
): Promise<MessageNode[]> {
  const messages = await invoke<RawMessageNode[]>("list_messages", {
    id,
    beforeOrderKey: toOptionalValue(beforeOrderKey),
    limit: toOptionalValue(limit)
  });
  return messages.map(fromRawMessageNode);
}

/**
 * 按需获取单个版本的完整正文。
 */
export async function getVersionContent(versionId: string): Promise<VersionContent> {
  const content = await invoke<RawVersionContent>("get_version_content", { versionId });
  return fromRawVersionContent(content);
}

/**
 * 切换某个楼层的 active version。
 */
export async function setActiveVersion(
  conversationId: string,
  nodeId: string,
  versionId: string
): Promise<void> {
  await invoke("set_active_version", {
    id: conversationId,
    nodeId,
    versionId
  });
}

/**
 * 删除指定版本。
 */
export async function deleteVersion(
  conversationId: string,
  nodeId: string,
  versionId: string
): Promise<DeleteVersionResult> {
  const result = await invoke<RawDeleteVersionResult>("delete_version", {
    id: conversationId,
    nodeId,
    versionId
  });
  return fromRawDeleteVersionResult(result);
}

/**
 * 发送消息并建立本次生成的事件通道。
 */
export async function sendMessage(
  id: string,
  input: SendMessageInput,
  onEvent?: (event: GenerationEvent) => void
): Promise<SendMessageResponse> {
  const eventChannel = createGenerationChannel(onEvent);
  const response = await invoke<RawDryRunResult | RawStartedResult>("send_message", {
    id,
    input: toRawSendMessageInput(input),
    eventChannel
  });
  return fromRawSendMessageResponse(response);
}

/**
 * 对指定楼层执行 reroll，并建立本次生成的事件通道。
 */
export async function reroll(
  id: string,
  nodeId: string,
  input?: RerollInput,
  onEvent?: (event: GenerationEvent) => void
): Promise<RerollResult> {
  const eventChannel = createGenerationChannel(onEvent);
  const result = await invoke<RawRerollResult>("reroll", {
    id,
    nodeId,
    input: toRawRerollInput(input),
    eventChannel
  });
  return fromRawRerollResult(result);
}

/**
 * 取消某个生成中的版本。
 */
export async function cancelGeneration(versionId: string): Promise<void> {
  await invoke("cancel_generation", { versionId });
}

export type {
  DeleteVersionResult,
  DryRunResult,
  GenerationEvent,
  MessageNode,
  MessageVersion,
  PromptMessage,
  RerollInput,
  RerollResult,
  SendMessageInput,
  SendMessageResponse,
  SendMessageResult,
  VersionContent
} from "./message-types";
export { toAppError, type AppError };
